File: BilisPera/BilisPera/Product/AlertView/BPAddressPickerView.m
//
//  BPAddressPickerView.m
//  BilisPera
//
//  Created by BHJ on 2025/8/25.
//

#import "BPAddressPickerView.h"

NS_ASSUME_NONNULL_BEGIN
@implementation BPAddressPickerViewModel

- (instancetype)init
{
    self = [super init];
    if (self) {
        self.contentHeight = kRatio(200);
    }
    return self;
}

@end

@interface BPAddressPickerView ()<UIPickerViewDelegate,UIPickerViewDataSource>

@property (nonatomic, strong) UIPickerView *pickerView;
@property (nonatomic, strong) UIButton *firstButton;
@property (nonatomic, strong) UIButton *secondButton;
@property (nonatomic, strong) UIButton *thirdButton;
@property (nonatomic, strong) UIView *lineView;
@property (nonatomic, strong) UIButton *selectedButton;
@property (nonatomic, strong) NSLayoutConstraint *indicatorCenterXConstraint;

@property (nonatomic, assign) NSInteger selectedType;
@property (nonatomic, strong) BPAddressModel *selectedProvince;
@property (nonatomic, strong) BPAddressModel *selectedCity;
@property (nonatomic, strong) BPAddressModel *selectedStreet;

@end

@implementation BPAddressPickerView

- (instancetype)initWithFrame:(CGRect)frame
{
    self = [super initWithFrame:frame];
    if (self) {
        [self setupUI];
        [self applyModel];
    }
    return self;
}

- (instancetype)initWithFrame:(CGRect)frame model:(BPAddressPickerViewModel *)model {
    self = [super initWithFrame:frame];
    if (self) {
        self.model = model;
        [self setupUI];
        [self applyModel];
    }
    return self;
}

- (void)setModel:(BPAddressPickerViewModel *)model {
    _model = model;
    [self applyModel];
}

#pragma mark - UI Setup

- (void)setupUI {
    self.backgroundColor = [UIColor clearColor];
    
    // PickerView
    _pickerView = [[UIPickerView alloc] initWithFrame:CGRectMake(0,
                                                                 0,
                                                                 kScreenW,
                                                                 kRatio(200))];
    _pickerView.backgroundColor = [UIColor clearColor];
    _pickerView.delegate = self;
    _pickerView.dataSource = self;
    _pickerView.translatesAutoresizingMaskIntoConstraints = NO;
    [self addSubview:_pickerView];
    
    // Buttons
    _firstButton = [self createButtonWithTitle:@"Street" tag:1000];
    _secondButton = [self createButtonWithTitle:@"Town" tag:1001];
    _thirdButton = [self createButtonWithTitle:@"State" tag:1002];
    
    // Line View
    _lineView = [UIView new];
    _lineView.backgroundColor = [UIColor blackColor];
    _lineView.translatesAutoresizingMaskIntoConstraints = NO;
    
    [self addSubview:_firstButton];
    [self addSubview:_secondButton];
    [self addSubview:_thirdButton];
    [self addSubview:_lineView];
    
    // Constraints
    [self setupConstraints];
}

- (UIButton *)createButtonWithTitle:(NSString *)title tag:(NSInteger)tag {
    UIButton *button = [UIButton buttonWithType:UIButtonTypeCustom];
    button.backgroundColor = [UIColor clearColor];
    [button setTitle:title forState:UIControlStateNormal];
    [button setTitleColor:[UIColor grayColor] forState:UIControlStateNormal];
    button.titleLabel.font = [UIFont systemFontOfSize:16];
    [button addTarget:self action:@selector(segementEvent:) forControlEvents:UIControlEventTouchUpInside];
    button.translatesAutoresizingMaskIntoConstraints = NO;
    button.tag = tag;
    return button;
}

- (void)setupConstraints {
    CGFloat buttonWidth = (kScreenW - kRatio(44)) / 3;
    self.indicatorCenterXConstraint = [self.lineView.centerXAnchor constraintEqualToAnchor:self.firstButton.centerXAnchor];
    
    [NSLayoutConstraint activateConstraints:@[
        // PickerView constraints
        [self.pickerView.leadingAnchor constraintEqualToAnchor:self.leadingAnchor],
        [self.pickerView.topAnchor constraintEqualToAnchor:self.topAnchor constant:kRatio(44)],
        [self.pickerView.trailingAnchor constraintEqualToAnchor:self.trailingAnchor],
        [self.pickerView.bottomAnchor constraintEqualToAnchor:self.bottomAnchor],
        [self.pickerView.heightAnchor constraintEqualToConstant:kRatio(436)],
        
        // First button constraints
        [self.firstButton.leadingAnchor constraintEqualToAnchor:self.leadingAnchor constant:kRatio(22)],
        [self.firstButton.topAnchor constraintEqualToAnchor:self.topAnchor],
        [self.firstButton.heightAnchor constraintEqualToConstant:kRatio(44)],
        [self.firstButton.widthAnchor constraintEqualToConstant:buttonWidth],
        
        // Second button constraints
        [self.secondButton.centerXAnchor constraintEqualToAnchor:self.centerXAnchor],
        [self.secondButton.topAnchor constraintEqualToAnchor:self.topAnchor],
        [self.secondButton.heightAnchor constraintEqualToConstant:kRatio(44)],
        [self.secondButton.widthAnchor constraintEqualToConstant:buttonWidth],
        
        // Third button constraints
        [self.thirdButton.trailingAnchor constraintEqualToAnchor:self.trailingAnchor constant:-kRatio(22)],
        [self.thirdButton.topAnchor constraintEqualToAnchor:self.topAnchor],
        [self.thirdButton.heightAnchor constraintEqualToConstant:kRatio(44)],
        [self.thirdButton.widthAnchor constraintEqualToConstant:buttonWidth],
        
        // Line view constraints
        self.indicatorCenterXConstraint,
        [self.lineView.topAnchor constraintEqualToAnchor:self.firstButton.bottomAnchor constant:-kRatio(5)],
        [self.lineView.heightAnchor constraintEqualToConstant:kRatio(3)],
        [self.lineView.widthAnchor constraintEqualToConstant:kRatio(32)]
    ]];
}

#pragma mark - Model Application

- (void)applyModel {
    [self.pickerView reloadAllComponents];
    [self segementEvent:self.firstButton];
}

#pragma mark - Button Actions

- (void)segementEvent:(UIButton *)sender {
    [sender setTitleColor:[UIColor blackColor] forState:UIControlStateNormal];
    [self.selectedButton setTitleColor:[UIColor grayColor] forState:UIControlStateNormal];
    self.selectedButton = sender;
    
    self.indicatorCenterXConstraint.active = NO;
    self.indicatorCenterXConstraint = [self.lineView.centerXAnchor constraintEqualToAnchor:sender.centerXAnchor];
    self.indicatorCenterXConstraint.active = YES;
    
    NSInteger tag = sender.tag - 1000;
    self.selectedType = tag;
    [self.pickerView reloadComponent:0];
    [self pickerView:self.pickerView didSelectRow:0 inComponent:0];
    
    [self setNeedsLayout];
}

- (void)nextStep {
    if (self.selectedProvince && !self.selectedCity) {
        [self segementEvent:self.secondButton];
    } else if (self.selectedProvince && self.selectedCity) {
        [self segementEvent:self.thirdButton];
    }
}

#pragma mark - UIPickerViewDataSource

- (NSInteger)numberOfComponentsInPickerView:(UIPickerView *)pickerView {
    return 1;
}

- (NSInteger)pickerView:(UIPickerView *)pickerView numberOfRowsInComponent:(NSInteger)component {
    switch (self.selectedType) {
        case 0: return self.model.dataSource.count;
        case 1: return self.selectedProvince.andwalked.count;
        case 2: return self.selectedCity.andwalked.count;
        default: return 0;
    }
}

#pragma mark - UIPickerViewDelegate

- (CGFloat)pickerView:(UIPickerView *)pickerView rowHeightForComponent:(NSInteger)component {
    return kRatio(50);
}

- (nullable NSString *)pickerView:(UIPickerView *)pickerView titleForRow:(NSInteger)row forComponent:(NSInteger)component {
    switch (self.selectedType) {
        case 0:
            return row < self.model.dataSource.count ? self.model
                .dataSource[row].tongues : nil;
        case 1:
            return row < self.selectedProvince.andwalked.count ? self.selectedProvince
                .andwalked[row].tongues : nil;
        case 2:
            return row < self.selectedCity.andwalked.count ? self.selectedCity
                .andwalked[row].tongues : nil;
        default:
            return nil;
    }
}

- (void)pickerView:(UIPickerView *)pickerView didSelectRow:(NSInteger)row inComponent:(NSInteger)component {
    NSString *address = @"";
    BOOL completed = NO;
    
    switch (self.selectedType) {
        case 0:
            if (row < self.model.dataSource.count) {
                self.selectedProvince = self.model.dataSource[row];
                address = self.selectedProvince.tongues ?: @"";
            }
            break;
            
        case 1:
            if (row < self.selectedProvince.andwalked.count) {
                self.selectedCity = self.selectedProvince.andwalked[row];
                NSString *province = self.selectedProvince.tongues ?: @"";
                NSString *city = self.selectedCity.tongues ?: @"";
                address = [NSString stringWithFormat:@"%@-%@", province, city];
            }
            break;
            
        case 2:
            if (row < self.selectedCity.andwalked.count) {
                self.selectedStreet = self.selectedCity.andwalked[row];
                NSString *province = self.selectedProvince.tongues ?: @"";
                NSString *city = self.selectedCity.tongues ?: @"";
                NSString *street = self.selectedStreet.tongues ?: @"";
                address = [NSString stringWithFormat:@"%@-%@-%@",
                           province,
                           city,
                           street];
                completed = YES;
            }
            break;
    }
    if (self.model.valueChanged) {
        self.model.valueChanged(address);
    }
    if (self.model.completed) {
        self.model.completed(completed);
    }
}

@end

NS_ASSUME_NONNULL_END

File: BilisPera/BilisPera/Product/AlertView/BPAddressPickerView.swift
//
//  BPAddressPickerView.swift
//  BilisPera
//
//  Created by BHJ on 2025/8/19.
//

import UIKit

class BPAddressPickerView: UIView {
    
    override init(frame: CGRect) {
        super.init(frame: frame)
        setupUI()
        applyModel()
    }
    
    init(frame: CGRect,model: BPAddressPickerViewModel = .init()) {
        self.model = model
        super.init(frame: frame)
        setupUI()
        applyModel()
    }
    
    required init?(coder: NSCoder) {
        fatalError("init(coder:) has not been implemented")
    }
    
    @objc var model: BPAddressPickerViewModel = .init(){
        didSet{
            applyModel()
        }
    }
    
    var selectedButton: UIButton?
    var indicatorCenterXConstraint: NSLayoutConstraint?
    
    lazy var pickerView : UIPickerView = {
        let view = UIPickerView.init(
            frame: CGRect.init(x: 0, y: 0, width: kScreenW, height: 200.ratio())
        )
        view.backgroundColor = UIColor.clear
        view.delegate = self
        view.dataSource = self
        view.translatesAutoresizingMaskIntoConstraints = false
        return view
    }()
    
    lazy var firstButton : UIButton = {
        let view = UIButton(type: .custom)
        view.backgroundColor = UIColor.clear
        view.setTitle("Street", for: .normal)
        view.setTitleColor(.gray, for: .normal)
        view.titleLabel?.font = UIFont.systemFont(ofSize: 16)
        view
            .addTarget(
                self,
                action: #selector(segementEvent(sender:)),
                for: .touchUpInside
            )
        view.translatesAutoresizingMaskIntoConstraints = false
        view.tag = 1000;
        return view
    }()
    
    lazy var secondButton : UIButton = {
        let view = UIButton(type: .custom)
        view.backgroundColor = UIColor.clear
        view.setTitle("Town", for: .normal)
        view.setTitleColor(.gray, for: .normal)
        view.titleLabel?.font = UIFont.systemFont(ofSize: 16)
        view
            .addTarget(
                self,
                action: #selector(segementEvent(sender:)),
                for: .touchUpInside
            )
        view.translatesAutoresizingMaskIntoConstraints = false
        view.tag = 1001;
        return view
    }()
    
    lazy var thirdButton : UIButton = {
        let view = UIButton(type: .custom)
        view.backgroundColor = UIColor.clear
        view.setTitle("State", for: .normal)
        view.setTitleColor(.gray, for: .normal)
        view.titleLabel?.font = UIFont.systemFont(ofSize: 16)
        view
            .addTarget(
                self,
                action: #selector(segementEvent(sender:)),
                for: .touchUpInside
            )
        view.translatesAutoresizingMaskIntoConstraints = false
        view.tag = 1002;
        return view
    }()
    
    lazy var lineView: UIView = {
        let view = UIView()
        view.backgroundColor = .black
        view.translatesAutoresizingMaskIntoConstraints = false
        return view
    }()
    
    private var selectedType: Int = 0
    private var selectedProvince: BPAddressModel?
    private var selectedCity: BPAddressModel?
    private var selectedStreet: BPAddressModel?
}

extension BPAddressPickerView: UIPickerViewDelegate,UIPickerViewDataSource{
    func numberOfComponents(in pickerView: UIPickerView) -> Int {
        return 1
    }
    
    func pickerView(_ pickerView: UIPickerView, numberOfRowsInComponent component: Int) -> Int {
        if selectedType == 0{
            return model.dataSource.count
        }else if selectedType == 1{
            return selectedProvince?.andwalked.count ?? 0
        }else{
            return selectedCity?.andwalked.count ?? 0
        }
    }
    
    func pickerView(_ pickerView: UIPickerView, rowHeightForComponent component: Int) -> CGFloat {
        return 50.ratio()
    }
    
    func pickerView(_ pickerView: UIPickerView, titleForRow row: Int, forComponent component: Int) -> String? {
        var title: String?
        if selectedType == 0{
            title = model.dataSource[row].tongues
        }else if selectedType == 1,selectedProvince?.andwalked.count ?? 0 > row{
            title = selectedProvince?.andwalked[row].tongues
        }else if selectedCity?.andwalked.count ?? 0 > row{
            title = selectedCity?.andwalked[row].tongues
        }
        return title
    }
    
    func pickerView(
        _ pickerView: UIPickerView,
        didSelectRow row: Int,
        inComponent component: Int
    ) {
        if selectedType == 0,model.dataSource.count > row{
            selectedProvince = model.dataSource[row]
            let address = selectedProvince?.tongues ?? ""
            model.valueChanged?(address)
            model.completed?(false)
        }else if selectedType == 1,selectedProvince?.andwalked.count ?? 0 > row{
            selectedCity = selectedProvince?.andwalked[row]
            let address = (selectedProvince?.tongues ?? "")  + "-" + (
                selectedCity?.tongues ?? ""
            )
            model.valueChanged?(address)
            model.completed?(false)
        }else if selectedCity?.andwalked.count ?? 0 > row{
            selectedStreet = selectedCity?.andwalked[row]
            let provice = (selectedProvince?.tongues ?? "")
            let cityName = (selectedCity?.tongues ?? "")
            let street = (selectedStreet?.tongues ?? "")
            let address = provice + "-" + cityName + "-" + street
            model.valueChanged?(address)
            model.completed?(true)
        }
    }
}

extension BPAddressPickerView{
    func setupUI(){
        backgroundColor = .clear
        addSubview(pickerView)
        NSLayoutConstraint.activate([
            pickerView.leadingAnchor.constraint(equalTo: leadingAnchor),
            pickerView.topAnchor
                .constraint(equalTo: topAnchor,constant: 44.ratio()),
            pickerView.trailingAnchor.constraint(equalTo: trailingAnchor),
            pickerView.bottomAnchor.constraint(equalTo: bottomAnchor),
            pickerView.heightAnchor.constraint(equalToConstant: 436.ratio()),
        ])
        addSubview(firstButton)
        addSubview(secondButton)
        addSubview(thirdButton)
        addSubview(lineView)
        self.indicatorCenterXConstraint = lineView.centerXAnchor
            .constraint(equalTo: firstButton.centerXAnchor)
        let buttonWidth = (kScreenW - 44.ratio()) / 3
        if let indiactorConstraint = self.indicatorCenterXConstraint {
            NSLayoutConstraint.activate(
                [
                    firstButton.leadingAnchor
                        .constraint(
                            equalTo: leadingAnchor,
                            constant: 22.ratio()
                        ),
                    firstButton.topAnchor.constraint(equalTo: topAnchor),
                    firstButton.heightAnchor
                        .constraint(equalToConstant: 44.ratio()),
                    firstButton.widthAnchor
                        .constraint(equalToConstant: buttonWidth),
                    
                    secondButton.centerXAnchor
                        .constraint(equalTo: centerXAnchor),
                    secondButton.topAnchor.constraint(equalTo: topAnchor),
                    secondButton.heightAnchor
                        .constraint(equalToConstant: 44.ratio()),
                    secondButton.widthAnchor
                        .constraint(equalToConstant: buttonWidth),

                    thirdButton.trailingAnchor
                        .constraint(
                            equalTo: trailingAnchor,
                            constant: -22.ratio()
                        ),
                    thirdButton.topAnchor.constraint(equalTo: topAnchor),
                    thirdButton.heightAnchor
                        .constraint(equalToConstant: 44.ratio()),
                    thirdButton.widthAnchor
                        .constraint(equalToConstant: buttonWidth),
                
                    indiactorConstraint,
                    lineView.topAnchor
                        .constraint(
                            equalTo: firstButton.bottomAnchor,
                            constant: -5.ratio()
                        ),
                    lineView.heightAnchor
                        .constraint(equalToConstant: 3.ratio()),
                    lineView.widthAnchor
                        .constraint(equalToConstant: 32.ratio()),
                ]
            )
        }
    }
    
    func applyModel(){
        pickerView.reloadAllComponents()
        self.segementEvent(sender: firstButton)
    }
    
    @objc func segementEvent(sender: UIButton) -> Void{
        sender.setTitleColor(.black, for: .normal);
        self.selectedButton?.setTitleColor(.gray, for: .normal)
        self.selectedButton = sender
        self.indicatorCenterXConstraint?.isActive = false
        self.indicatorCenterXConstraint =  lineView.centerXAnchor
            .constraint(equalTo: sender.centerXAnchor)
        self.indicatorCenterXConstraint?.isActive = true
        let tag = sender.tag - 1000
        self.selectedType = tag;
        self.pickerView.reloadComponent(0)
        self.pickerView(pickerView, didSelectRow: 0, inComponent: 0);
        setNeedsLayout()
    }
    
    @objc func nextStep() -> Void{
        if ((self.selectedProvince != nil) && (self.selectedCity == nil)) {
            self.segementEvent(sender: secondButton)
        }else if((self.selectedProvince != nil) && (self.selectedCity != nil)){
            self.segementEvent(sender: thirdButton)
        }
    }
}

@objc
extension BPAddressPickerView{
    @MainActor
    @objc class BPAddressPickerViewModel: NSObject {
        @objc var contentHeight = 200.ratio()
        @objc var dataSource: [BPAddressModel] = []
        @objc var valueChanged: ((String)-> Void)? = nil
        @objc var completed: ((Bool)-> Void)? = nil
    }
}
